clients/aws-sdk-bedrock-runtime/tests/integration/test_bidirectional_streaming.py

```
@@ -0,0 +1,280 @@
+    chunk_count = 0
+    with AUDIO_FILE.open("rb") as f:
+        while chunk := f.read(CHUNK_SIZE):
+            chunk_count += 1
+            encoded_chunk = base64.b64encode(chunk).decode("utf-8")
+            await _send_event(
+                stream,
+                AUDIO_INPUT_EVENT % (prompt_name, audio_content_name, encoded_chunk),
+            )
+            # 512 bytes / (16000 Hz * 2 bytes/sample) = 0.016s per chunk
+            await asyncio.sleep(0.016)
+
+    assert chunk_count > 0, f"No audio chunks were sent from {AUDIO_FILE}"
+
+    silence_chunk = bytes(CHUNK_SIZE)
+    encoded_silence = base64.b64encode(silence_chunk).decode("utf-8")
+    for _ in range(SILENCE_CHUNKS):
+        await _send_event(
+            stream,
+            AUDIO_INPUT_EVENT % (prompt_name, audio_content_name, encoded_silence),
+        )
+        await asyncio.sleep(0.016)
+
+    await _send_event(stream, CONTENT_END_EVENT % (prompt_name, audio_content_name))
+    await asyncio.sleep(RESPONSE_WAIT_TIME)
+    await _send_event(stream, PROMPT_END_EVENT % prompt_name)
+    await _send_event(stream, SESSION_END_EVENT)
+
+
+async def _receive_stream_output(
+    stream: DuplexEventStream[
+        InvokeModelWithBidirectionalStreamInput,
+        InvokeModelWithBidirectionalStreamOutput,
+        InvokeModelWithBidirectionalStreamOperationOutput,
+    ],
+) -> tuple[bool, bool, list[str]]:
+    """Receive and collect output from the bidirectional stream.
+
+    Returns:
+        Tuple of (got_text, got_audio, all_text_output)
+    """
+    got_text = False
+    got_audio = False
+    all_text_output: list[str] = []
+
+    await stream.await_output()
+    output_stream = stream.output_stream
+    if output_stream is None:
+        return got_text, got_audio, all_text_output
+
+    async for out in output_stream:
+        if not isinstance(out, InvokeModelWithBidirectionalStreamOutputChunk):
+            continue
```

What else are we receiving in the stream that we're skipping here?

---

The stream can yield both normal OutputChunk events and modeled error events (InternalServerException, ValidationException, ThrottlingException, etc.). Those error types don't have a bytes_ payload, so the isinstance check narrows the type and keeps pyright happy.

---

Can this be successful then if we get an error chunk as the last part of our stream and just ignore it? I don't immediately remember how termination of the stream works.

---

Good catch. You're right that if we receive error events after getting some successful chunks the test would still pass since we just continue and the assertions only check that we got some output. Should we fail explicitly on error events? or is it acceptable for the stream to send partial results followed by an error?

---

A successful stream shouldn't terminate with errors so I think we'd want to consider that a failure case. I know there have been a lot of unintuitive behaviors with the Bedrock APIs and how they surface errors. I would not be surprised if we get back partial data before it fails, so we should account for that in the test.

If we're seeing typing issues, let's take a look at that. I wonder if we're missing partial definitions or type unions somewhere.